eos-phone-home

```diff
@@ -0,0 +1,406 @@
+
+        request_data = json.dumps(request)
+        print("Request data:", request_data)
+
+        print("Sending to server (%s)..." % endpoint)
+
+        if self._debug:
+            print("Debugging turned on so nothing sent to server!")
+            return True
+
+        success = False
+
+        try:
+            data = request_data.encode('utf-8')
+
+            req = urllib.request.Request(endpoint, data=data, method='PUT')
+            req.add_header('Content-Type', 'application/json')
+
+            response_data = None
+            with urllib.request.urlopen(req) as conn:
+                response_data = conn.read().decode('utf-8')
+
+            print('Response:', response_data)
+
+            response = json.loads(response_data)
+
+            if 'success' in response and \
+               response['success'] == True:
+                print("Server returned success message!")
+                success = True
+            else:
+                print("Server failed to process the data:", response)
+
+        except Exception as e:
+            print("Sending failed!", e, file=sys.stderr)
+
+        return success
+
+    def _need_to_activate(self):
+        return not os.path.exists(os.path.join(self.STATE_DIRECTORY,
+                                               'activated'))
+
+    def _do_activate(self):
+        if self._send_to_server(self.ACTIVATION_ENDPOINT,
+                                self.ACTIVATION_VARIABLES):
+            # We're done, mark us activated
+            with open(os.path.join(self.STATE_DIRECTORY, 'activated'), 'w'):
+                pass
+
+            print("Activated!")
+        else:
+            print("Could not send activation data to server!", file=sys.stderr)
+
+            return False
+
+        return True
+
+    def _need_to_ping(self):
+        if self._lookup_or_get_variable('live'):
+            print("Not sending ping from live system.")
+            return False
+
+        if not os.path.exists(os.path.join(self.STATE_DIRECTORY, 'count')):
+            print("Creating new count file!")
+            self._set_count(0)
+            return True
+
+        count_time = os.path.getmtime(os.path.join(self.STATE_DIRECTORY,
+                                                   'count'))
+        count_age = time.time() - count_time
+        print("Count age:", count_age)
+
+        if count_age < 0:
+            print("Clock has gone backwards? Resetting time so we will ping",
+              "again in 24 hours.")
+
+            count = self._lookup_or_get_variable('count')
+            self._set_count(count)
+
+            return False
+
+        # we aim to ping daily
+        DAY_IN_SECONDS = 60 * 60 * 24
+        return count_age >= DAY_IN_SECONDS
+
+    def _do_ping(self):
+        if self._send_to_server(self.PING_ENDPOINT, self.PING_VARIABLES):
+            # increment the counter upon successful ping only
+            count = self._lookup_or_get_variable('count')
```

Comment:
Not entirely happy with this because the property of the original system was that by incrementing the count only on successful ping, the sum of all distinct counters (provided you do freq[count-1]--; freq[count]++ each time you get a ping in) approximates the number of unique systems. If we remove that property and have "free wheeling" counters, we're relying on either daily connectivity (which is a bad idea) or we have to use some other probabilistic heuristics to guess at how many distinct systems we have, so we lower the quality of our data significantly.

Obviously we have to do some heuristic decaying (ie we don't count non-movers, eventually - but over a longer period) but I'd prioritise more accurate system counting over more accurate uptime tracking of an indistinct number of users.

Canonical actually later added a "first run" date to their parameters, which means they can compare ping count to time elapsed, possibly for the reasons you described, but I elected not to include that because my "gut feeling" was that it would de-anonymise the information too much and could be seen as a trackable attribute.